southbound/device-factory/device-factory-core/src/test/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/RecordHandlingTest.java

```diff
@@ -310,4 +334,94 @@ void testVariableService() throws Exception {
+
+        // Auto-compute timestamp format: nanoseconds
+        bulks.clear();
+        config.mapping.put("@timestamp", "timestamp_ns");
+        deviceMapper.handle(config, Map.of(), new byte[0]);
+        dto = getResourceValue("provider", "data", "val");
+        assertEquals(now, dto.timestamp);
+    }
+
+    @Test
+    void testDateTime() throws Exception {
+        final DeviceMappingConfigurationDTO config = prepareConfig();
+
+        // Set a record
+        final Instant now = Instant.now();
+        final OffsetDateTime date = now.atOffset(ZoneOffset.ofHours(1));
+        final DateTimeFormatter formatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG, FormatStyle.MEDIUM);
+        final String strPattern = "MM dd, yyyy - Q - hh:mm a - s.n - X";
+
+        final Map<String, Object> record = new HashMap<>();
+        record.put("provider", "provider");
+        record.put("iso", date.toString());
+        record.put("custom", date.format(DateTimeFormatter.ofPattern(strPattern)));
+        record.put("locale_fr", date.format(formatter.withLocale(Locale.FRANCE)));
+        record.put("locale_de", date.format(formatter.withLocale(Locale.GERMAN)));
+        record.put("locale_en", date.format(formatter.withLocale(Locale.UK)));
+        record.put("locale_zh", date.format(formatter.withLocale(Locale.SIMPLIFIED_CHINESE)));
+        parser.setRecords(record);
+
+        // ISO format
+        config.mapping.put("@provider", "provider");
+        config.mapping.put("data/val", null);
+        config.mapping.put("@datetime", "iso");
+        config.mappingOptions.formatDateTime = null;
+        config.mappingOptions.dateTimezone = null;
+        deviceMapper.handle(config, Map.of(), new byte[0]);
+        GenericDto dto = getResourceValue("provider", "data", "val");
+        assertEquals(now, dto.timestamp);
+
+        // Custom pattern
+        bulks.clear();
+        config.mapping.put("@datetime", "custom");
+        config.mappingOptions.formatDateTime = strPattern;
+        deviceMapper.handle(config, Map.of(), new byte[0]);
+        dto = getResourceValue("provider", "data", "val");
+        assertEquals(now, dto.timestamp);
+
+        // Locale-based date (mix of country and language codes)
+        for (Locale locale : List.of(Locale.FRANCE, Locale.GERMAN, Locale.UK, Locale.SIMPLIFIED_CHINESE)) {
+            bulks.clear();
+            config.mapping.put("@datetime", "locale_" + locale.getLanguage());
+            config.mappingOptions.formatDateTime = null;
+            config.mappingOptions.formatDateTimeLocale = locale.toString();
+            config.mappingOptions.formatDateStyle = "long";
+            config.mappingOptions.formatTimeStyle = "medium";
+            config.mappingOptions.dateTimezone = "+01";
+            deviceMapper.handle(config, Map.of(), new byte[0]);
+            dto = getResourceValue("provider", "data", "val");
+            assertEquals(now.truncatedTo(ChronoUnit.SECONDS), dto.timestamp);
```

Comment:
I think that this is vulnerable to rollover failures.
```suggestion
            long delta = Math.abs(ChronoUnit.SECONDS.between(Instant.now(), dto.timestamp));
            assertTrue(delta < 2, "The dto timestamp " + dto.timestamp + " is skewed from " + Instant.now());
```

Reply:
Not sure on that one: we're comparing the same instant

Reply:
Ok - I wasn't sure if we had some auto-generated timestamps. If it's actually the same time being used then the existing check is great.